src/scf/direct_minimization.jl

```diff
@@ -0,0 +1,132 @@
+# Direct minimization of the energy
+
+using Optim
+
+# This is all a bit annoying because our Psi is represented as Psi[k][G,n], and Optim accepts only dense arrays
+# We do a bit of back and forth using custom `vec` (ours -> optim's) and `devec` (optim's -> ours) functions
+
+# Impose the orthogonality constraint on orbitals inside each k-block
+struct DMManifold <: Optim.Manifold
+    Nk::Int
+    devec_fun::Function
+end
+function Optim.project_tangent!(m::DMManifold, g, x)
+    g_devec = m.devec_fun(g)
+    x_devec = m.devec_fun(x)
+    for ik = 1:m.Nk
+        Optim.project_tangent!(Optim.Stiefel(),
+                               g_devec[ik],
+                               x_devec[ik])
+    end
+    g
+end
+function Optim.retract!(m::DMManifold, x)
+    x_devec = m.devec_fun(x)
+    for ik = 1:m.Nk
+        Optim.retract!(Optim.Stiefel(), x_devec[ik])
+    end
+    x
+end
+
+# Array of preconditioners
+struct DMPreconditioner
+    Nk::Int
+    Pks::Array # Pks[ik] is the preconditioner for kpoint ik
+    devec_fun::Function
+end
+function LinearAlgebra.ldiv!(p, P::DMPreconditioner, d)
+    p_devec = P.devec_fun(p)
+    d_devec = P.devec_fun(d)
+    for ik = 1:P.Nk
+        ldiv!(p_devec[ik], P.Pks[ik], d_devec[ik])
+    end
+    p
+end
+function LinearAlgebra.dot(x, P::DMPreconditioner, y)
+    x_devec = P.devec_fun(x)
+    y_devec = P.devec_fun(y)
+    sum(dot(x_devec[ik], P.Pks[ik], y_devec[ik])
+        for ik = 1:P.Nk)
+end
+function Optim.precondprep!(P::DMPreconditioner, x)
+    x_devec = P.devec
+    for ik = 1:P.Nk
+        precondprep!(Pks[ik], x_devec[ik])
+    end
+    P
+end
+
+
+"""
+Computes the ground state by direct minimization. `kwargs...` are passed to `Optim.Options()`.
+"""
+direct_minimization(basis::PlaneWaveBasis; kwargs...) = direct_minimization(basis, nothing; kwargs...)
+function direct_minimization(basis::PlaneWaveBasis{T}, Psi0;
+                             prec_type=PreconditionerTPA,
+                             optim_solver=Optim.LBFGS, kwargs...) where T
```

Comment:
General question: Does direct minimization not work with smearing, or would that just complicate things?

Reply:
It complicates things quite a lot, because then you need to optimize the free energy wrt the occupations as well as the orbitals.

Reply:
There's a few papers on this, but nothing really fully satisfactory - that's one of the points that would be great to investigate in the future

Reply:
Ok. Was just wondering why there was no comment or mention about this.